fix: safari in-app notification template glitches

LetItRock opened the pull request:
- **What kind of change does this PR introduce?** (Bug fix, feature, docs update, ...)
The In-App Notification Template Editor: fixes the issue with the notification content field glitches on Safari and allows for multiline content.

- **What is the current behavior?** (You can also link to an open issue here)
[Issue link](https://github.com/novuhq/novu/issues/914)
The In-App Notification Template Editor: on Safari when typing anything to the notification content field it has glitches. The cursor moves on every character type to the beginning of the field.

- **What is the new behavior (if this is a feature change)?**
Users can type anything in the notification content field without any glitches/issues. Multiline content is possible.

- **Screenshots**:
![Screenshot 2022-08-01 at 23 02 37](https://user-images.githubusercontent.com/2607232/182481618-bbbad02d-f156-4cea-bbb3-c245cecdde65.png)

https://user-images.githubusercontent.com/2607232/182481636-cf19ab90-5d0f-454b-b276-d4ece87875cf.mov

https://user-images.githubusercontent.com/2607232/182481672-609c5a7c-5a0d-4719-89f4-8c05c1a2db78.mov

LetItRock commented on apps/web/cypress/support/commands.ts:
```diff
@@ -17,6 +17,15 @@ Cypress.Commands.add('seed', () => {
   return cy.request('POST', `${Cypress.env('apiUrl')}/v1/testing/seed`, {});
 });
 
+Cypress.Commands.add('eventsTrigger', (body: Cypress.RequestBody, apiKey: string) => {
+  return cy.request({
+    method: 'POST',
+    url: `${Cypress.env('apiUrl')}/v1/events/trigger`,
+    body,
+    headers: { 'Content-Type': 'application/json', Authorization: `ApiKey ${apiKey}` },
+  });
+});
```
Added this command that allows triggering the event. I do use it in the test below.

scopsy replied:
This is a great idea for reuse! We already have a `createNotification` task in cypress that does something similar a use of it can be found in the activity page tests:
```
cy.initializeSession()
      .as('session')
      .then((session: any) => {
        cy.wait(500);

        return cy.task('createNotifications', {
          identifier: session.templates[0].triggers[0].identifier,
          token: session.token,
          count: 25,
          organizationId: session.organization._id,
          environmentId: session.environment._id,
        });
      });
```

Do you think we need another one as command here?

LetItRock replied:
ohh cool! I was looking for that by didn't found, I'll try to reuse that ;)

LetItRock commented on apps/web/cypress/tests/notifications-editor.spec.ts:
```diff
@@ -120,6 +120,51 @@ describe('Notifications Creator', function () {
       cy.getByTestId('trigger-snippet-btn').click();
       cy.location('pathname').should('equal', '/templates');
     });
+    it('should create multiline in-app notification, send it and receive', function () {
```
create an in-app notification with multiline content, trigger it and then verify if it's rendered correctly on the notifications list

scopsy replied:
A great idea for a test!

LetItRock commented on apps/web/cypress/tests/notifications-editor.spec.ts:
```diff
@@ -538,7 +583,7 @@ describe('Notifications Creator', function () {
       });
     });
 
-    it.only('should be able to delete a step', function () {
```
because of this `.only`, other tests were not running... I've removed that `.only` and run all of them locally, they were passing so I've decided to leave that...

scopsy replied:
Oh great catch! @novuhq/novu-team check it out aswell, we should try to pay a really close attention to this.

scopsy replied:
@LetItRock created #947 to make sure this won't happen in the future too

LetItRock commented on apps/web/cypress/tests/notifications-editor.spec.ts:
```diff
@@ -266,7 +311,7 @@ describe('Notifications Creator', function () {
     });
 
-    it('should edit notification', function () {
+    it.skip('should edit notification', function () {
```
this test was breaking, I'm not sure why... check the comment below

scopsy replied:
I will take a look at it once merge, do you mind opening an issue for that?

LetItRock replied:
sure will do that!

LetItRock replied:
@scopsy the ticket https://github.com/novuhq/novu/issues/948

LetItRock commented on apps/web/src/components/templates/in-app-editor/ContentContainer.tsx:
```diff
@@ -17,30 +18,11 @@ export function ContentContainer({
 }) {
   const theme = useMantineTheme();
   const ref = useRef<HTMLDivElement>(null);
-  const [visiblePlaceholder, setVisiblePlaceholder] = useState(!!value);
```
There is no need for additional `useState`. This component is controlled meaning when value is passed it will re-render, so we can check that `value` and show/hide placeholder

scopsy replied:
Great idea, love it!

LetItRock commented on apps/web/src/components/templates/in-app-editor/ContentContainer.tsx:
```diff
@@ -17,30 +18,11 @@ export function ContentContainer({
 
-  const handleChange = (e: React.ChangeEvent<HTMLDivElement>) => {
-    const textContent = e.target.textContent;
-    if (ref?.current) {
-      ref.current.textContent = textContent;
-    }
```
this line was causing the issue, setting the `textContent` removes the `the node's children` and replaces it with a new `text node`

LetItRock commented on apps/web/src/components/templates/in-app-editor/ContentContainer.tsx:
```diff
@@ -17,30 +18,11 @@ export function ContentContainer({
-    onChange(textContent);
-  };
-
-  useEffect(() => {
```
no need to do that

scopsy replied:
This defiently make sense here 🙏

LetItRock commented on apps/web/src/components/templates/in-app-editor/ContentContainer.tsx:
```diff
@@ -17,30 +18,11 @@ export function ContentContainer({
-
   useEffect(() => {
-    if (value !== ref.current?.textContent) {
+    if (ref.current && value !== ref.current?.innerHTML) {
```
this effect won't trigger after the `onChange` event, because the `value` and `innerHTML` will be the same

LetItRock commented on apps/web/src/components/templates/in-app-editor/ContentContainer.tsx:
```diff
@@ -54,7 +36,7 @@ export function ContentContainer({
           __html: content as string,
         }}
-        onKeyUp={(e: any) => handleChange(e)}
+        onKeyUp={(e: any) => onChange(e.target.innerHTML)}
```
only the `innerHTML` value works well with `dangerouslySetInnerHTML` for multiline content

LetItRock commented on packages/notification-center/src/components/notification-bell/NotificationBell.tsx:
```diff
@@ -26,7 +26,7 @@ export function NotificationBell(props: INotificationBellProps) {
 
   return (
-    <ActionIcon variant="transparent">
+    <ActionIcon variant="transparent" data-test-id="notification-bell">
```
used in e2e test

scopsy commented on apps/web/cypress/tests/notifications-editor.spec.ts:
```diff
@@ -120,6 +120,51 @@ describe('Notifications Creator', function () {
+      cy.getByTestId('trigger-snippet-btn').click();
+
+      // trigger the notification
+      cy.eventsTrigger(
+        {
+          name: 'test-notification-title',
+          to: {
+            subscriberId: this.session.user.id,
+            email: this.session.user.email,
+          },
+          payload: {},
+        },
+        this.session.environment.apiKeys[0].key
+      ).then(() => {
```
Following up my previous comment this can be used as:

```
 cy.task('createNotifications', {
    identifier: 'test-notification-title',
    token: this.session.environment.apiKeys[0].key,
    count: 1,
    organizationId: this.session.organization._id,
    environmentId: this.session.environment._id,
  });

```

Do you think it will be  ok here?

scopsy reviewed (commented):
Thank you so much @LetItRock for you contribution. The really enjoyed going over your PR and loved the way you communicated the thought lines using comments here. 

I will try to look if the edit test also fails on `main` branch just to make sure it's not affected by this change.

LetItRock commented:
> Thank you so much @LetItRock for you contribution. The really enjoyed going over your PR and loved the way you communicated the thought lines using comments here.
> 
> I will try to look if the edit test also fails on `main` branch just to make sure it's not affected by this change.

@scopsy hey, thanks! 
yeah, comments are a great way of communicating the intention. It's always better to leave a few in the PR's to reduce the amount of questions.

Sure, let me know if anything will be needed to fix if that test will fail.

LetItRock commented:
@scopsy hey, I was able to reuse the `cy.task('createNotifications'...` 🎉  but needed to slightly modify it by adding the `subscriberId` as the arg. Also removed that command that I've added.
Please take a look when you have time.

scopsy reviewed (approved):
Amazing @LetItRock Merging this once tests are passing
